text/0000-crate-quarantine.md

```diff
@@ -0,0 +1,156 @@
+
+So far, crates.io has been relatively fortunate to only have isolated incidents[^rustdecimal], but the increasing popularity of Rust only makes it more and more likely that we will also be the target of similar spam and malware attacks.
+
+There are mitigating factors at play: in particular, our reliance on GitHub for authentication, while problematic from purist FOSS and ecosystem diversity perspectives, means that we benefit from GitHub's machinery in terms of detecting bad actors.
+
+Regardless, we need to be able to prevent malicious packages from being uploaded, both in emergency situations (for example, mitigating spam attacks) and in more mundane cases where we can confidently identify a potentially malicious user. This RFC proposes an off-by-default quarantine approach that allows us to examine packages before they are published when it is considered warranted.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This project will add a new admin feature to crates.io that allows crate publishes matching specific properties to be placed into a quarantine state instead of being immediately published.
+
+A version in the quarantined state will have its `.crate` file published to S3 to simplify review, and will appear in crates.io API responses as an unpublished crate, but will _not_ be added to the index — and will therefore _not_ be installable via `cargo` — until it has been reviewed and approved.
+
+`cargo publish` will also be modified to report when an uploaded crate version has been placed into quarantine. 
+
+Once a crate version has been quarantined, then a member of the crates.io team will review the quarantined crate within 1 week[^sla]. Once reviewed, the crate may be either approved and published to the index, or rejected and deleted. Either way, the crate owner(s) will be notified by e-mail, assuming crates.io has a verified e-mail available for their account.
+
+Should the crate owner(s) disagree with the quarantine action, they may appeal to the moderation team.
+
+## When will a quarantine rule be added?
+
+Ultimately, this will be decided by the crates.io team on a case by case basis. There are two primary scenarios that are currently envisioned where this functionality would be used:
+
+### Large scale spam attack
+
+In the event of a large scale spam attack on the entire site, we may want to either restrict individual users (if the spam is coming from a finite set of users) or quarantine all published versions temporarily (using a `.*` crate name rule) until it subsides, at which point we can analyse which crate versions are actually valid.
+
+### Malicious users
+
+When we have high confidence (due to user reports or automated scanning) that a user is uploading malicious crates then a rule can be added for that specific user until the crates.io team is able to communicate with them.
```

Comment:
This phrasing implies that you will always use this mechanism, rather than (for instance) banning a malicious user. You may not want to pre-commit to that as the required action; doing so would remove flexibility.

Reply:
Definitely not the intent, thanks! Slight clarification applied.